### PR TITLE
Feat/gh 314/more shared widgets tests

### PR DESCRIPTION
**Pull request by andre-tm-hui**

### CODE REVIEW

File: test/presentation/shared_widgets/no_ripple_behavior_test.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      );
+      await tester.pumpAndSettle();
+
+      expect(find.byKey(key), findsOneWidget);
+      expect(
+          tester
+              .firstWidget<ScrollConfiguration>(find.byKey(key))
+              .behavior
+              .runtimeType,
+          NoRippleBehavior);
```
Comment (Xazin) [inline test/presentation/shared_widgets/no_ripple_behavior_test.dart:26]: I would love a comma here, but looks good haha

File: test/presentation/shared_widgets/participant_avatars_test.dart [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  });
+  group('ParticipantAvatars tests:', () {
+    testWidgets('can render initial state', (WidgetTester tester) async {
+      when(() => topParticipantsBloc.state)
+          .thenAnswer((_) => TopParticipantsState.initial());
+
+      await buildAndPump(
+          tester: tester,
+          widget:
+              TopParticipantAvatars(crowdActionId: tCrowdactionNoPassword.id));
```
Comment (Xazin) [inline test/presentation/shared_widgets/participant_avatars_test.dart:40]: I want more commas hahaha

File: test/presentation/shared_widgets/participant_avatars_test.dart [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    });
+
+    testWidgets('can render fetching state', (WidgetTester tester) async {
+      when(() => topParticipantsBloc.state)
+          .thenAnswer((_) => TopParticipantsState.fetching());
+
+      await buildAndPump(
+          tester: tester,
+          widget:
+              TopParticipantAvatars(crowdActionId: tCrowdactionNoPassword.id));
```
Comment (Xazin) [inline test/presentation/shared_widgets/participant_avatars_test.dart:53]: Commas :D

File: test/presentation/shared_widgets/participant_avatars_test.dart [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      await tester.pump();
+
+      expect(find.byType(TopParticipantsShimmer), findsOneWidget);
+    });
+
+    testWidgets('fetched 1 participants', (WidgetTester tester) async {
+      when(() => topParticipantsBloc.state).thenAnswer(
+          (_) => TopParticipantsState.fetched(List.filled(1, tParticipation)));
```
Comment (Xazin) [inline test/presentation/shared_widgets/participant_avatars_test.dart:61]: List.filled seems to be a bit overkill when you can just do `[tParticipation]`

File: test/presentation/shared_widgets/participant_avatars_test.dart [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+      await buildAndPump(
+          tester: tester,
+          widget:
+              TopParticipantAvatars(crowdActionId: tCrowdactionNoPassword.id));
```
Comment (Xazin) [inline test/presentation/shared_widgets/participant_avatars_test.dart:66]: Commas :D

File: test/presentation/shared_widgets/participant_avatars_test.dart [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    });
+
+    testWidgets('fetched 3 participants', (WidgetTester tester) async {
+      when(() => topParticipantsBloc.state).thenAnswer(
+          (_) => TopParticipantsState.fetched(List.filled(3, tParticipation)));
+
+      await buildAndPump(
+          tester: tester,
+          widget:
+              TopParticipantAvatars(crowdActionId: tCrowdactionNoPassword.id));
```
Comment (Xazin) [inline test/presentation/shared_widgets/participant_avatars_test.dart:80]: Commas

File: test/presentation/shared_widgets/participant_avatars_test.dart [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    testWidgets('fetched any other number of participants',
+        (WidgetTester tester) async {
+      when(() => topParticipantsBloc.state).thenAnswer(
+          (_) => TopParticipantsState.fetched(List.filled(10, tParticipation)));
+
+      await buildAndPump(
+          tester: tester,
+          widget:
+              TopParticipantAvatars(crowdActionId: tCrowdactionNoPassword.id));
```
Comment (Xazin) [inline test/presentation/shared_widgets/participant_avatars_test.dart:95]: Commas :D

File: test/presentation/shared_widgets/phone_input_test.dart [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+void main() {
+  const channel = MethodChannel('com.julienvignali/phone_number');
+
+  setUp(() async {
+    channel.setMockMethodCallHandler((MethodCall methodCall) async {
+      if (methodCall.arguments['string'] == '31 0612345678') {
+        return {'isValid': true};
+      }
+      return null;
+    });
```
Comment (Xazin) [inline test/presentation/shared_widgets/phone_input_test.dart:16]:
```suggestion
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return methodCall.arguments['string'] == '31 0612345678' ?  {'isValid': true} : null;
    });
```

File: test/presentation/shared_widgets/phone_input_test.dart [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  });
+
+  tearDown(() {
+    channel.setMockMethodCallHandler(null);
+  });
+
+  group('PhoneInput tests:', () {
+    testWidgets('can render', (WidgetTester tester) async {
+      await buildAndPump(
+          tester: tester, widget: PhoneInput(TextEditingController()));
```
Comment (Xazin) [inline test/presentation/shared_widgets/phone_input_test.dart:26]: Commas :D

File: test/presentation/shared_widgets/phone_input_test.dart [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+      expect(find.byType(PhoneInput), findsOneWidget);
+      expect(find.text('Enter a valid phone number'), findsOneWidget);
+      expect(find.text('+31'), findsOneWidget);
+    });
+
+    testWidgets('correct message when number is invalid',
+        (WidgetTester tester) async {
+      await buildAndPump(
+          tester: tester, widget: PhoneInput(TextEditingController()));
```
Comment (Xazin) [inline test/presentation/shared_widgets/phone_input_test.dart:37]: Commas :D

File: test/presentation/shared_widgets/phone_input_test.dart [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+      await buildAndPump(
+          tester: tester,
+          widget: PhoneInput(
+            TextEditingController(),
+            isValid: (_) {
+              isValid = true;
+            },
+            onChange: (_) {
+              onChange = true;
+            },
```
Comment (Xazin) [inline test/presentation/shared_widgets/phone_input_test.dart:73]:
```suggestion
            isValid: (_) => isValid = true,
            onChange: (_) => onChange = true,
```

File: test/presentation/shared_widgets/photo_selector_test.dart [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    });
+  });
+}
+
+Future<XFile?> getFileFromAssets(String filename) async {
+  final byteData = await rootBundle.load('assets/images/$filename');
+
+  final file = File('${Directory.systemTemp.path}/$filename');
+  await file.writeAsBytes(byteData.buffer
+      .asUint8List(byteData.offsetInBytes, byteData.lengthInBytes));
```
Comment (Xazin) [inline test/presentation/shared_widgets/photo_selector_test.dart:92]: Commas :D

File: test/presentation/shared_widgets/photo_selector_test.dart [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+  XFile? xFile = XFile(file.path);
+
+  return xFile;
```
Comment (Xazin) [inline test/presentation/shared_widgets/photo_selector_test.dart:96]:
```suggestion
  return XFile(file.path);
```

File: test/presentation/shared_widgets/pill_button_test.dart [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+    testWidgets('callback calls when enabled', (WidgetTester tester) async {
+      bool onTapResult = false;
+      await buildAndPump(
+          tester: tester,
+          widget: PillButton(
+              text: 'text',
+              onTap: () {
+                onTapResult = true;
+              }));
```
Comment (Xazin) [inline test/presentation/shared_widgets/pill_button_test.dart:23]:
```suggestion
      await buildAndPump(
          tester: tester,
          widget: PillButton(
              text: 'text',
              onTap: () => onTapResult = true,
          ),
      );
```

File: test/presentation/shared_widgets/pill_button_test.dart [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+      bool onTapResult = false;
+      await buildAndPump(
+          tester: tester,
+          widget: PillButton(
+            text: 'text',
+            onTap: () {
+              onTapResult = true;
+            },
+            isLoading: true,
+          ));
```
Comment (Xazin) [inline test/presentation/shared_widgets/pill_button_test.dart:58]: Same as the other suggestions

File: test/presentation/shared_widgets/rectangle_button_test.dart [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+      await tester.pumpAndSettle();
+
+      expect(find.byType(RectangleButton), findsOneWidget);
+      expect(find.text('text'), findsOneWidget);
+    });
+
+    testWidgets('callback calls on tap when enabled',
+        (WidgetTester tester) async {
+      bool onTapResponse = false;
+      await buildAndPump(
```
Comment (Xazin) [inline test/presentation/shared_widgets/rectangle_button_test.dart:19]: Same as the other suggestions

File: test/presentation/shared_widgets/rectangle_button_test.dart [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+          tester: tester,
+          widget: RectangleButton(
+            text: 'text',
+            onTap: () {
+              onTapResponse = true;
+            },
+            enabled: false,
+          ));
```
Comment (Xazin) [inline test/presentation/shared_widgets/rectangle_button_test.dart:44]: Same as the other suggestions

File: test/presentation/shared_widgets/web_view_page_test.dart [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+import 'package:webview_flutter/webview_flutter.dart';
+
+import '../../test_helper.dart';
+
+void main() {
+  group('WebViewPage tests:', () {
+    testWidgets('can render', (WidgetTester tester) async {
+      await buildAndPump(
+          tester: tester,
+          widget: WebViewPage(url: 'https://collaction.org', title: 'title'));
```
Comment (Xazin) [inline test/presentation/shared_widgets/web_view_page_test.dart:12]: Commas :D

File: test/presentation/shared_widgets/web_view_page_test.dart [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+      await tester.pumpAndSettle();
+
+      expect(find.byType(WebViewPage), findsOneWidget);
+      expect(find.text('title'), findsOneWidget);
+    });
+
+    testWidgets('links the correct webpage', (WidgetTester tester) async {
+      await buildAndPump(
+          tester: tester,
+          widget: WebViewPage(url: 'https://collaction.org', title: 'title'));
```
Comment (Xazin) [inline test/presentation/shared_widgets/web_view_page_test.dart:22]: Commas :D

File: test/presentation/shared_widgets/web_view_page_test.dart [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+      await tester.pumpAndSettle();
+
+      expect(find.byType(WebView), findsOneWidget);
+      expect(tester.firstWidget<WebView>(find.byType(WebView)).initialUrl,
+          equals('https://collaction.org'));
```
Comment (Xazin) [inline test/presentation/shared_widgets/web_view_page_test.dart:27]: Commas :D

File: test/presentation/shared_widgets/no_ripple_behavior_test.dart (hunk 1 above)
Comment (andre-tm-hui) [reply test/presentation/shared_widgets/no_ripple_behavior_test.dart]: Did not realize commas made it look so much better haha. I've gone through and added them wherever appropriate

File: test/presentation/shared_widgets/participant_avatars_test.dart (hunk 4 above)
Comment (andre-tm-hui) [reply test/presentation/shared_widgets/participant_avatars_test.dart]: Changed, just had it there to match the subsequent tests

File: test/presentation/shared_widgets/phone_input_test.dart (hunk 8 above)
Comment (andre-tm-hui) [reply test/presentation/shared_widgets/phone_input_test.dart]: Works no problem, updated now

File: test/presentation/shared_widgets/phone_input_test.dart (hunk 11 above)
Comment (andre-tm-hui) [reply test/presentation/shared_widgets/phone_input_test.dart]: Changed wherever applicable

File: test/presentation/shared_widgets/photo_selector_test.dart (hunk 13 above)
Comment (andre-tm-hui) [reply test/presentation/shared_widgets/photo_selector_test.dart]: Fixed

**Review by Xazin — APPROVED** [review]

**Pull request merged**